// typesxml/dist/index.js
"use strict";
/*******************************************************************************
 * Copyright (c) 2023 - 2024 Maxprograms.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse   License 1.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/epl-v10.html
 *
 * Contributors:
 *     Maxprograms - initial API and implementation
 *******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Grammar = exports.ContentModel = exports.NotationDecl = exports.InternalSubset = exports.EntityDecl = exports.ElementDecl = exports.DTDParser = exports.AttListDecl = exports.AttDecl = exports.XMLWriter = exports.XMLUtils = exports.XMLElement = exports.XMLDocumentType = exports.XMLDocument = exports.XMLDeclaration = exports.XMLComment = exports.XMLAttribute = exports.TextNode = exports.SAXParser = exports.ProcessingInstruction = exports.Indenter = exports.FileReader = exports.DOMBuilder = exports.Constants = exports.Catalog = exports.CData = void 0;
var CData_1 = require("./CData");
Object.defineProperty(exports, "CData", { enumerable: true, get: function () { return CData_1.CData; } });
var Catalog_1 = require("./Catalog");
Object.defineProperty(exports, "Catalog", { enumerable: true, get: function () { return Catalog_1.Catalog; } });
var Constants_1 = require("./Constants");
Object.defineProperty(exports, "Constants", { enumerable: true, get: function () { return Constants_1.Constants; } });
var DOMBuilder_1 = require("./DOMBuilder");
Object.defineProperty(exports, "DOMBuilder", { enumerable: true, get: function () { return DOMBuilder_1.DOMBuilder; } });
var FileReader_1 = require("./FileReader");
Object.defineProperty(exports, "FileReader", { enumerable: true, get: function () { return FileReader_1.FileReader; } });
var Indenter_1 = require("./Indenter");
Object.defineProperty(exports, "Indenter", { enumerable: true, get: function () { return Indenter_1.Indenter; } });
var ProcessingInstruction_1 = require("./ProcessingInstruction");
Object.defineProperty(exports, "ProcessingInstruction", { enumerable: true, get: function () { return ProcessingInstruction_1.ProcessingInstruction; } });
var SAXParser_1 = require("./SAXParser");
Object.defineProperty(exports, "SAXParser", { enumerable: true, get: function () { return SAXParser_1.SAXParser; } });
var TextNode_1 = require("./TextNode");
Object.defineProperty(exports, "TextNode", { enumerable: true, get: function () { return TextNode_1.TextNode; } });
var XMLAttribute_1 = require("./XMLAttribute");
Object.defineProperty(exports, "XMLAttribute", { enumerable: true, get: function () { return XMLAttribute_1.XMLAttribute; } });
var XMLComment_1 = require("./XMLComment");
Object.defineProperty(exports, "XMLComment", { enumerable: true, get: function () { return XMLComment_1.XMLComment; } });
var XMLDeclaration_1 = require("./XMLDeclaration");
Object.defineProperty(exports, "XMLDeclaration", { enumerable: true, get: function () { return XMLDeclaration_1.XMLDeclaration; } });
var XMLDocument_1 = require("./XMLDocument");
Object.defineProperty(exports, "XMLDocument", { enumerable: true, get: function () { return XMLDocument_1.XMLDocument; } });
var XMLDocumentType_1 = require("./XMLDocumentType");
Object.defineProperty(exports, "XMLDocumentType", { enumerable: true, get: function () { return XMLDocumentType_1.XMLDocumentType; } });
var XMLElement_1 = require("./XMLElement");
Object.defineProperty(exports, "XMLElement", { enumerable: true, get: function () { return XMLElement_1.XMLElement; } });
var XMLUtils_1 = require("./XMLUtils");
Object.defineProperty(exports, "XMLUtils", { enumerable: true, get: function () { return XMLUtils_1.XMLUtils; } });
var XMLWriter_1 = require("./XMLWriter");
Object.defineProperty(exports, "XMLWriter", { enumerable: true, get: function () { return XMLWriter_1.XMLWriter; } });
var AttDecl_1 = require("./dtd/AttDecl");
Object.defineProperty(exports, "AttDecl", { enumerable: true, get: function () { return AttDecl_1.AttDecl; } });
var AttListDecl_1 = require("./dtd/AttListDecl");
Object.defineProperty(exports, "AttListDecl", { enumerable: true, get: function () { return AttListDecl_1.AttListDecl; } });
var DTDParser_1 = require("./dtd/DTDParser");
Object.defineProperty(exports, "DTDParser", { enumerable: true, get: function () { return DTDParser_1.DTDParser; } });
var ElementDecl_1 = require("./dtd/ElementDecl");
Object.defineProperty(exports, "ElementDecl", { enumerable: true, get: function () { return ElementDecl_1.ElementDecl; } });
var EntityDecl_1 = require("./dtd/EntityDecl");
Object.defineProperty(exports, "EntityDecl", { enumerable: true, get: function () { return EntityDecl_1.EntityDecl; } });
var InternalSubset_1 = require("./dtd/InternalSubset");
Object.defineProperty(exports, "InternalSubset", { enumerable: true, get: function () { return InternalSubset_1.InternalSubset; } });
var NotationDecl_1 = require("./dtd/NotationDecl");
Object.defineProperty(exports, "NotationDecl", { enumerable: true, get: function () { return NotationDecl_1.NotationDecl; } });
var ContentModel_1 = require("./grammar/ContentModel");
Object.defineProperty(exports, "ContentModel", { enumerable: true, get: function () { return ContentModel_1.ContentModel; } });
var Grammar_1 = require("./grammar/Grammar");
Object.defineProperty(exports, "Grammar", { enumerable: true, get: function () { return Grammar_1.Grammar; } });
