// typesxml/dist/dtd/ElementDecl.js
"use strict";
/*******************************************************************************
 * Copyright (c) 2023 - 2024 Maxprograms.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse   License 1.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/epl-v10.html
 *
 * Contributors:
 *     Maxprograms - initial API and implementation
 *******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElementDecl = void 0;
const Constants_1 = require("../Constants");
class ElementDecl {
    name;
    contentSpec;
    constructor(name, contentSpec) {
        this.name = name;
        this.contentSpec = contentSpec;
    }
    getName() {
        return this.name;
    }
    getContentSpec() {
        return this.contentSpec;
    }
    getNodeType() {
        return Constants_1.Constants.ELEMENT_DECL_NODE;
    }
    toString() {
        return '<!ELEMENT ' + this.name + ' ' + this.contentSpec + '>';
    }
    equals(node) {
        // TODO
        throw new Error("Method not implemented.");
    }
}
exports.ElementDecl = ElementDecl;
