// typesxml/dist/dtd/InternalSubset.js
"use strict";
/*******************************************************************************
 * Copyright (c) 2023 - 2024 Maxprograms.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse   License 1.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/epl-v10.html
 *
 * Contributors:
 *     Maxprograms - initial API and implementation
 *******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalSubset = void 0;
const Constants_1 = require("../Constants");
const DTDParser_1 = require("./DTDParser");
class InternalSubset {
    declarationText;
    grammar;
    constructor(declaration) {
        this.declarationText = declaration;
        let parser = new DTDParser_1.DTDParser();
        this.grammar = parser.parseString(declaration.substring(1, declaration.length - 1));
    }
    getNodeType() {
        return Constants_1.Constants.INTERNAL_SUBSET_NODE;
    }
    toString() {
        return this.declarationText;
    }
    equals(node) {
        // TODO Auto-generated method stub
        return false;
    }
}
exports.InternalSubset = InternalSubset;
