// typesxml/dist/dtd/NotationDecl.js
"use strict";
/*******************************************************************************
 * Copyright (c) 2023 - 2024 Maxprograms.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse   License 1.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/epl-v10.html
 *
 * Contributors:
 *     Maxprograms - initial API and implementation
 *******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotationDecl = void 0;
const Constants_1 = require("../Constants");
class NotationDecl {
    name;
    publicId;
    systemId;
    constructor(name, publicId, systemId) {
        this.name = name;
        this.publicId = publicId;
        this.systemId = systemId;
    }
    getName() {
        return this.name;
    }
    getNodeType() {
        return Constants_1.Constants.NOTATION_DECL_NODE;
    }
    toString() {
        // TODO
        throw new Error("Method not implemented.");
    }
    equals(node) {
        // TODO
        throw new Error("Method not implemented.");
    }
}
exports.NotationDecl = NotationDecl;
